### PR TITLE
fix: aggregate JSON output parameters correctly

### LINKED ISSUE #13510: JSON script `result` vs JSON output parameters are treated differently for `withParam` item aggregation
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened? What did you expect to happen?

For a workflow where we want to have a nested fan-out, i.e. process a list-of-lists, using the `.result` of a script template is treated differently from using a named output parameter when collecting results, meaning we _must_ use `stdout` script results to process a list-of-lists (which is not ideal).

This is due to the controller unmarshalling individual items of the results list here https://github.com/argoproj/argo-workflows/blob/f8f1893799b67446a833897af7f860fa08e71b1b/workflow/controller/operator.go#L3244

And then re-marshalling the whole results list here https://github.com/argoproj/argo-workflows/blob/f8f1893799b67446a833897af7f860fa08e71b1b/workflow/controller/operator.go#L3252

Which leads to a correct input parameter passed to the inner DAG: 
image

Individual output parameters of the fan-out are not unmarshalled during processing here https://github.com/argoproj/argo-workflows/blob/f8f1893799b67446a833897af7f860fa08e71b1b/workflow/controller/operator.go#L3232-L3239

But _are_ re-marshalled here https://github.com/argoproj/argo-workflows/blob/f8f1893799b67446a833897af7f860fa08e71b1b/workflow/controller/operator.go#L3259

Which leads to a serialized JSON string (with quotes escaped) passed to the inner DAG:
image

Resulting in the error message
```
withParam value could not be parsed as a JSON list: [{\"key1\": \"value1\"}, {\"key2\": \"value2\"}, {\"key3\": \"value3\"}]: invalid character '\\' looking for beginning of object key string
```

Investigated and confirmed with @Joibel, who tried patching the controller and got the output parameters version working, by attempting to unmarshal individual output parameters to then re-marshal them (and falling back to existing mechanism if it fails) - PR from Alan will be incoming 😄 

### Version(s)

v3.5.10

### Paste a minimal workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: double-fan-out-using-result-
spec:
  entrypoint: main-dag
  templates:
    - inputs:
        parameters:
          - name: param1
      name: operation
      outputs:
        parameters:
          - name: output1
            valueFrom:
              path: /tmp/fan_out.json
      script:
        command:
          - python3
        image: python:3.11
        source: |-
          import os
          import sys
          sys.path.append(os.getcwd())
          import json
          try: param1 = json.loads(r'''{{inputs.parameters.param1}}''')
          except: param1 = r'''{{inputs.parameters.param1}}'''

          with open('/tmp/fan_out.json', 'w') as f:
              json.dump(param1, f)
          print(json.dumps(param1))
    - dag:
        tasks:
          - arguments:
              parameters:
                - name: param1
                  value: '{{item}}'
            name: task3
            template: operation
            withParam: '{{inputs.parameters.param1}}'
      inputs:
        parameters:
          - name: param1
      name: secondary-dag
    - dag:
        tasks:
          - arguments:
              parameters:
                - name: param1
                  value: '[[{"key1": "value1"}, {"key2": "value2"}, {"key3": "value3"}], [{"key4": "value4"}, {"key5": "value5"}]]'
            name: task1
            template: operation
          - arguments:
              parameters:
                - name: param1
                  value: '{{item}}'
            depends: task1
            name: task2
            template: operation
            withParam: '{{tasks.task1.outputs.result}}'
          - arguments:
              parameters:
                - name: param1
                  value: '{{item}}'
            depends: task2
            name: task3-dag
            template: secondary-dag
            withParam: '{{tasks.task2.outputs.result}}'
      name: main-dag
---
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: double-fan-out-using-param-
spec:
  entrypoint: main-dag
  templates:
    - inputs:
        parameters:
          - name: param1
      name: operation
      outputs:
        parameters:
          - name: output1
            valueFrom:
              path: /tmp/fan_out.json
      script:
        command:
          - python3
        image: python:3.11
        source: |-
          import os
          import sys
          sys.path.append(os.getcwd())
          import json
          try: param1 = json.loads(r'''{{inputs.parameters.param1}}''')
          except: param1 = r'''{{inputs.parameters.param1}}'''

          with open('/tmp/fan_out.json', 'w') as f:
              json.dump(param1, f)
          print(json.dumps(param1))
    - dag:
        tasks:
          - arguments:
              parameters:
                - name: param1
                  value: '{{item}}'
            name: task3
            template: operation
            withParam: '{{inputs.parameters.param1}}'
      inputs:
        parameters:
          - name: param1
      name: secondary-dag
    - dag:
        tasks:
          - arguments:
              parameters:
                - name: param1
                  value: '[[{"key1": "value1"}, {"key2": "value2"}, {"key3": "value3"}], [{"key4": "value4"}, {"key5": "value5"}]]'
            name: task1
            template: operation
          - arguments:
              parameters:
                - name: param1
                  value: '{{item}}'
            depends: task1
            name: task2
            template: operation
            withParam: '{{tasks.task1.outputs.parameters.output1}}'
          - arguments:
              parameters:
                - name: param1
                  value: '{{item}}'
            depends: task2
            name: task3-dag
            template: secondary-dag
            withParam: '{{tasks.task2.outputs.parameters.output1}}'
      name: main-dag
```

### Logs from the workflow controller

```text
time="2024-08-27T10:02:37.702Z" level=info msg="Processing workflow" Phase= ResourceVersion=208400 namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:37.708Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=0 workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:37.708Z" level=info msg="Updated phase  -> Running" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:37.708Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:37.708Z" level=info msg="was unable to obtain node for , letting display name to be nodeName" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:37.708Z" level=info msg="DAG node double-fan-out-using-param-n2zxs initialized Running" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:37.708Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-3329134287, taskName task3-dag"
time="2024-08-27T10:02:37.708Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-2909490305, taskName task2"
time="2024-08-27T10:02:37.708Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-2859157448, taskName task1"
time="2024-08-27T10:02:37.708Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-2859157448, taskName task1"
time="2024-08-27T10:02:37.708Z" level=info msg="All of node double-fan-out-using-param-n2zxs.task1 dependencies [] completed" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:37.708Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:37.708Z" level=info msg="Pod node double-fan-out-using-param-n2zxs-2859157448 initialized Pending" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:37.713Z" level=info msg="Created pod: double-fan-out-using-param-n2zxs.task1 (double-fan-out-using-param-n2zxs-operation-2859157448)" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:37.713Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-2909490305, taskName task2"
time="2024-08-27T10:02:37.713Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-3329134287, taskName task3-dag"
time="2024-08-27T10:02:37.713Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-2909490305, taskName task2"
time="2024-08-27T10:02:37.713Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-3329134287, taskName task3-dag"
time="2024-08-27T10:02:37.713Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:37.713Z" level=info msg=reconcileAgentPod namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:37.722Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=208408 workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.713Z" level=info msg="Processing workflow" Phase=Running ResourceVersion=208408 namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.713Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=1 workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.713Z" level=info msg="task-result changed" namespace=argo nodeID=double-fan-out-using-param-n2zxs-2859157448 workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.713Z" level=info msg="node changed" namespace=argo new.message= new.phase=Succeeded new.progress=0/1 nodeID=double-fan-out-using-param-n2zxs-2859157448 old.message= old.phase=Pending old.progress=0/1 workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.713Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-3329134287, taskName task3-dag"
time="2024-08-27T10:02:47.713Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-2909490305, taskName task2"
time="2024-08-27T10:02:47.713Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-2909490305, taskName task2"
time="2024-08-27T10:02:47.714Z" level=info msg="TaskGroup node double-fan-out-using-param-n2zxs-2909490305 initialized Running (message: )" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.714Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-2735842384, taskName task2(0:[{\"key1\":\"value1\"} {\"key2\":\"value2\"} {\"key3\":\"value3\"}])"
time="2024-08-27T10:02:47.714Z" level=info msg="All of node double-fan-out-using-param-n2zxs.task2(0:[{\"key1\":\"value1\"} {\"key2\":\"value2\"} {\"key3\":\"value3\"}]) dependencies [task1] completed" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.714Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.714Z" level=info msg="Pod node double-fan-out-using-param-n2zxs-2735842384 initialized Pending" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.720Z" level=info msg="Created pod: double-fan-out-using-param-n2zxs.task2(0:[{\"key1\":\"value1\"} {\"key2\":\"value2\"} {\"key3\":\"value3\"}]) (double-fan-out-using-param-n2zxs-operation-2735842384)" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.720Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-1865176775, taskName task2(1:[{\"key4\":\"value4\"} {\"key5\":\"value5\"}])"
time="2024-08-27T10:02:47.720Z" level=info msg="All of node double-fan-out-using-param-n2zxs.task2(1:[{\"key4\":\"value4\"} {\"key5\":\"value5\"}]) dependencies [task1] completed" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.721Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.721Z" level=info msg="Pod node double-fan-out-using-param-n2zxs-1865176775 initialized Pending" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.727Z" level=info msg="Created pod: double-fan-out-using-param-n2zxs.task2(1:[{\"key4\":\"value4\"} {\"key5\":\"value5\"}]) (double-fan-out-using-param-n2zxs-operation-1865176775)" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.727Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-3329134287, taskName task3-dag"
time="2024-08-27T10:02:47.727Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-3329134287, taskName task3-dag"
time="2024-08-27T10:02:47.727Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.727Z" level=info msg=reconcileAgentPod namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.739Z" level=info msg="Workflow update successful" namespace=argo phase=Running resourceVersion=208475 workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:47.747Z" level=info msg="cleaning up pod" action=labelPodCompleted key=argo/double-fan-out-using-param-n2zxs-operation-2859157448/labelPodCompleted
time="2024-08-27T10:02:57.722Z" level=info msg="Processing workflow" Phase=Running ResourceVersion=208475 namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.723Z" level=info msg="Task-result reconciliation" namespace=argo numObjs=3 workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.723Z" level=info msg="task-result changed" namespace=argo nodeID=double-fan-out-using-param-n2zxs-2859157448 workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.723Z" level=info msg="task-result changed" namespace=argo nodeID=double-fan-out-using-param-n2zxs-1865176775 workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.723Z" level=info msg="task-result changed" namespace=argo nodeID=double-fan-out-using-param-n2zxs-2735842384 workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.723Z" level=info msg="node changed" namespace=argo new.message= new.phase=Succeeded new.progress=0/1 nodeID=double-fan-out-using-param-n2zxs-1865176775 old.message= old.phase=Pending old.progress=0/1 workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.723Z" level=info msg="node changed" namespace=argo new.message= new.phase=Succeeded new.progress=0/1 nodeID=double-fan-out-using-param-n2zxs-2735842384 old.message= old.phase=Pending old.progress=0/1 workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.724Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-3329134287, taskName task3-dag"
time="2024-08-27T10:02:57.724Z" level=info msg="node double-fan-out-using-param-n2zxs-2909490305 phase Running -> Succeeded" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.724Z" level=info msg="node double-fan-out-using-param-n2zxs-2909490305 finished: 2024-08-27 10:02:57.72440034 +0000 UTC" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.724Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-3329134287, taskName task3-dag"
time="2024-08-27T10:02:57.724Z" level=info msg="TaskGroup node double-fan-out-using-param-n2zxs-3329134287 initialized Running (message: )" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.724Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-109176038, taskName task3-dag(0:[{\\\"key1\\\": \\\"value1\\\"}, {\\\"key2\\\": \\\"value2\\\"}, {\\\"key3\\\": \\\"value3\\\"}])"
time="2024-08-27T10:02:57.724Z" level=info msg="All of node double-fan-out-using-param-n2zxs.task3-dag(0:[{\\\"key1\\\": \\\"value1\\\"}, {\\\"key2\\\": \\\"value2\\\"}, {\\\"key3\\\": \\\"value3\\\"}]) dependencies [task2] completed" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.724Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.724Z" level=info msg="DAG node double-fan-out-using-param-n2zxs-109176038 initialized Running" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.724Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-1233275690, taskName task3"
time="2024-08-27T10:02:57.724Z" level=info msg="Skipped node double-fan-out-using-param-n2zxs-1233275690 initialized Error (message: withParam value could not be parsed as a JSON list: [{\\\"key1\\\": \\\"value1\\\"}, {\\\"key2\\\": \\\"value2\\\"}, {\\\"key3\\\": \\\"value3\\\"}]: invalid character '\\\\' looking for beginning of object key string)" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.724Z" level=info msg="Outbound nodes of double-fan-out-using-param-n2zxs-109176038 set to [double-fan-out-using-param-n2zxs-1233275690]" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.724Z" level=info msg="node double-fan-out-using-param-n2zxs-109176038 phase Running -> Error" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.724Z" level=info msg="node double-fan-out-using-param-n2zxs-109176038 finished: 2024-08-27 10:02:57.724873674 +0000 UTC" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.724Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-2799279137, taskName task3-dag(1:[{\\\"key4\\\": \\\"value4\\\"}, {\\\"key5\\\": \\\"value5\\\"}])"
time="2024-08-27T10:02:57.724Z" level=info msg="All of node double-fan-out-using-param-n2zxs.task3-dag(1:[{\\\"key4\\\": \\\"value4\\\"}, {\\\"key5\\\": \\\"value5\\\"}]) dependencies [task2] completed" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.724Z" level=warning msg="Node was nil, will be initialized as type Skipped" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.725Z" level=info msg="DAG node double-fan-out-using-param-n2zxs-2799279137 initialized Running" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.725Z" level=warning msg="was unable to obtain the node for double-fan-out-using-param-n2zxs-3008809221, taskName task3"
time="2024-08-27T10:02:57.725Z" level=info msg="Skipped node double-fan-out-using-param-n2zxs-3008809221 initialized Error (message: withParam value could not be parsed as a JSON list: [{\\\"key4\\\": \\\"value4\\\"}, {\\\"key5\\\": \\\"value5\\\"}]: invalid character '\\\\' looking for beginning of object key string)" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.725Z" level=info msg="Outbound nodes of double-fan-out-using-param-n2zxs-2799279137 set to [double-fan-out-using-param-n2zxs-3008809221]" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.725Z" level=info msg="node double-fan-out-using-param-n2zxs-2799279137 phase Running -> Error" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.725Z" level=info msg="node double-fan-out-using-param-n2zxs-2799279137 finished: 2024-08-27 10:02:57.725148549 +0000 UTC" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.725Z" level=info msg="node double-fan-out-using-param-n2zxs-3329134287 phase Running -> Error" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.725Z" level=info msg="node double-fan-out-using-param-n2zxs-3329134287 finished: 2024-08-27 10:02:57.725200757 +0000 UTC" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.725Z" level=info msg="Outbound nodes of double-fan-out-using-param-n2zxs set to [double-fan-out-using-param-n2zxs-1233275690 double-fan-out-using-param-n2zxs-3008809221]" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.725Z" level=info msg="node double-fan-out-using-param-n2zxs phase Running -> Error" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.725Z" level=info msg="node double-fan-out-using-param-n2zxs finished: 2024-08-27 10:02:57.725286882 +0000 UTC" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.725Z" level=info msg="TaskSet Reconciliation" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.725Z" level=info msg=reconcileAgentPod namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.725Z" level=info msg="Updated phase Running -> Error" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.725Z" level=info msg="Marking workflow completed" namespace=argo workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.735Z" level=info msg="Workflow update successful" namespace=argo phase=Error resourceVersion=208577 workflow=double-fan-out-using-param-n2zxs
time="2024-08-27T10:02:57.803Z" level=info msg="cleaning up pod" action=labelPodCompleted key=argo/double-fan-out-using-param-n2zxs-operation-2735842384/labelPodCompleted
time="2024-08-27T10:02:57.803Z" level=info msg="cleaning up pod" action=labelPodCompleted key=argo/double-fan-out-using-param-n2zxs-operation-1865176775/labelPodCompleted
```

### Logs from in your workflow's wait container

```text
time="2024-08-27T10:02:50.896Z" level=info msg="Main container completed" error="<nil>"
time="2024-08-27T10:02:50.896Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
time="2024-08-27T10:02:50.896Z" level=info msg="Saving output parameters"
time="2024-08-27T10:02:50.896Z" level=info msg="Saving path output parameter: output1"
time="2024-08-27T10:02:50.896Z" level=info msg="Copying /tmp/fan_out.json from base image layer"
time="2024-08-27T10:02:50.896Z" level=info msg="Successfully saved output parameter: output1"
time="2024-08-27T10:02:50.896Z" level=info msg="No output artifacts"
time="2024-08-27T10:02:50.904Z" level=info msg="Alloc=9382 TotalAlloc=13577 Sys=24677 NumGC=3 Goroutines=8"
time="2024-08-27T10:02:50.909Z" level=info msg="Deadline monitor stopped"
time="2024-08-27T10:02:50.909Z" level=info msg="stopping progress monitor (context done)" error="context canceled"
time="2024-08-27T10:02:48.887Z" level=info msg="Executor initialized" deadline="0001-01-01 00:00:00 +0000 UTC" includeScriptOutput=false namespace=argo podName=double-fan-out-using-param-n2zxs-operation-2735842384 templateName=operation version="&Version{Version:v3.5.10,BuildDate:2024-08-01T05:12:26Z,GitCommit:25829927431d9a0f46d17b72ae74aedb8d700884,GitTag:v3.5.10,GitTreeState:clean,GoVersion:go1.21.12,Compiler:gc,Platform:linux/arm64,}"
time="2024-08-27T10:02:48.895Z" level=info msg="Starting deadline monitor"
time="2024-08-27T10:02:50.897Z" level=info msg="Main container completed" error="<nil>"
time="2024-08-27T10:02:50.897Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
time="2024-08-27T10:02:50.897Z" level=info msg="Saving output parameters"
time="2024-08-27T10:02:50.897Z" level=info msg="Saving path output parameter: output1"
time="2024-08-27T10:02:50.897Z" level=info msg="Copying /tmp/fan_out.json from base image layer"
time="2024-08-27T10:02:50.897Z" level=info msg="Successfully saved output parameter: output1"
time="2024-08-27T10:02:50.897Z" level=info msg="No output artifacts"
time="2024-08-27T10:02:50.904Z" level=info msg="Alloc=9269 TotalAlloc=13587 Sys=24165 NumGC=3 Goroutines=8"
time="2024-08-27T10:02:38.989Z" level=info msg="Starting deadline monitor"
time="2024-08-27T10:02:40.992Z" level=info msg="Main container completed" error="<nil>"
time="2024-08-27T10:02:40.992Z" level=info msg="No Script output reference in workflow. Capturing script output ignored"
time="2024-08-27T10:02:40.992Z" level=info msg="Saving output parameters"
time="2024-08-27T10:02:40.992Z" level=info msg="Saving path output parameter: output1"
time="2024-08-27T10:02:40.992Z" level=info msg="Copying /tmp/fan_out.json from base image layer"
time="2024-08-27T10:02:40.992Z" level=info msg="Successfully saved output parameter: output1"
time="2024-08-27T10:02:40.992Z" level=info msg="No output artifacts"
time="2024-08-27T10:02:41.007Z" level=info msg="Alloc=9787 TotalAlloc=13590 Sys=23909 NumGC=3 Goroutines=8"
time="2024-08-27T10:02:41.012Z" level=info msg="stopping progress monitor (context done)" error="context canceled"
```

### PR DESCRIPTION
**Pull request by Joibel**
### Motivation

Fixes #13510

### Modifications

`outputs.result` of a `withItems`/`withParams` will be conditionally `json.Marshal`ed.

Other output parameters are always json.Marshalled, which leads to inconsitency and a complete inability to pass JSON out of withItems outputs straight back in as `withItems`

Conditionally Marshal the outputs.params
* avoiding it if the parameter isn't enclosed in `{` or `[`, primarily to avoid bare numbers as they're valid JSON but need quoting here
* falling back to old behaviour if this fails

### Verification

Additional e2e test tests the example case.

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -3265,9 +3265,38 @@ func (woc *wfOperationCtx) processAggregateNodeOutputs(scope *wfScope, prefix st
...
+		for _, value := range valueList {
+			// Only try to unmarshal things that look like json lists or dicts
+			// and especially avoid unstringified numbers which are valid JSON
+			valueTrim := strings.Trim(value, " ")
+			valueTrimLen := len(valueTrim)
+			if valueTrimLen > 0 &&
+				!((valueTrim[0] == '{' && valueTrim[valueTrimLen-1] == '}') ||
+					(valueTrim[0] == '[' && valueTrim[valueTrimLen-1] == ']')) {
+				unmarshalSuccess = false
+				break // This isn't a json list or dict, leave it
```
[the remark this answers is not in the document]
Comment (isubasinghe) [reply workflow/controller/operator.go:3279]: Why not unmarshal and check the type?

[the remark this answers is not in the document]
Comment (Joibel) [reply workflow/controller/operator.go]: I've split it into two functions, and written (overlapping) unit tests for them. operator_test.go is unnecessarily huge, so I've made a new test file.

I've also implemented your unmarshal+type check as a much cleaner way of doing this - thanks for the suggestion. The string parsing is used elsewhere for JSON deliniation, but I prefer this.

**Comment by agilgur5** [comment]
@Joibel is this at all impacted by #12909? Since that creates a diff between `main` and `3.5` with JSON in templates


Briefly looking at the code, I don't think it does since it only impacts output parameters, but wanted to double check, especially since https://github.com/argoproj/argo-workflows/pull/12909#issuecomment-2285300935 is still incomplete

**Comment by Joibel** [comment]
> @Joibel is this at all impacted by #12909? Since that creates a diff between `main` and `3.5` with JSON in templates
> 
> Briefly looking at the code, I don't think it does since it only impacts output parameters, but wanted to double check, especially since [#12909 (comment)](https://github.com/argoproj/argo-workflows/pull/12909#issuecomment-2285300935) is still incomplete

No, this isn't overlapping, and wouldn't be if #12909 had the remaining changes which are mostly around docs.

#13532 also doesn't appear to overlap, but does seem to have a similar cause. I haven't had a chance to see what's happening there though.

**Comment by agilgur5** [comment]
> No, this isn't overlapping

Thanks just wanted to make sure!

> and wouldn't be if #12909 had the remaining changes which are mostly around docs.

There were some edge cases there that hadn't been tested (that I had asked tests to be added for), and if those edge cases are impacted, we might want to change the logic.

Do you want me to write up an issue to handle the remaining changes to track?

**Comment by agilgur5** [comment]
> Do you want me to write up an issue to handle the remaining changes to track? I was going to do so if it hadn't been handled soon so that we don't forget about it

Went ahead and created an issue for it: #13541

**Review by isubasinghe — APPROVED** [review]
LGTM
